staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -671,19 +744,41 @@ func (c *Cacher) processEvent(event *watchCacheEvent) {
 }
 
 func (c *Cacher) dispatchEvents() {
+	startID := bucketID(c.clock.Now())
 	for {
 		select {
 		case event, ok := <-c.incoming:
 			if !ok {
 				return
 			}
-			c.dispatchEvent(&event)
+			if event.Type == watch.Bookmark {
+				currentID := bucketID(c.clock.Now())
+				c.dispatchBookmarkEvent(&event, startID, currentID)
+				startID = currentID + 1
+			} else {
+				c.dispatchEvent(&event)
+			}
 		case <-c.stopCh:
 			return
 		}
 	}
 }
 
+func (c *Cacher) dispatchBookmarkEvent(event *watchCacheEvent, startID, currentID int64) {
+	c.Lock()
+	defer c.Unlock()
+	for id := startID; id <= currentID; id++ {
+		watchers := c.watchers.popWatchers(id)
+		for _, w := range watchers {
+			if w.stopped {
+				continue
+			}
+			w.nonblockingAdd(event)
+			c.watchers.pushToNextTimeBucket(w, currentID+1)
```

Comment:
Given how we are implementing it in the first version (i.e. timouet-2s), this will never be effective.
So instead of doing that, I would rather add a comment that once we decide to change the strategy to more sophisticated, we should also push add a watcher.